File: app/src/main/java/libertypassage/com/app/utilis/Utility.java
package libertypassage.com.app.utilis;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityManager;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.gson.Gson;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

import libertypassage.com.app.R;
import libertypassage.com.app.models.CountryDetail;
import libertypassage.com.app.models.DetailIndustryProf;
import libertypassage.com.app.models.DetailLocationHistory;
import libertypassage.com.app.models.ModelMyLocation;


public class Utility {
    private static String PREFERENCE = "LIBERTY";
    private static String KEY_COUNTRY_LIST = "country_list";
    private static String KEY_INDUSTRY_PROF = "industry_profession";
    private static String KEY_LOCATION_HISTORY = "location_history";
    private static ProgressDialog progressDialog;

    public static void setSharedPreference(Context context, String name, String value) {
        SharedPreferences settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = settings.edit();
        editor.putString(name, value);
        editor.apply();
    }

    public static String getSharedPreferences(Context context, String name) {
        SharedPreferences settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        return settings.getString(name, "");
    }

    public static void clearPreference(Context context) {
        SharedPreferences settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = settings.edit();
        editor.clear();
        editor.apply();
    }

    public static void showProgressDialog(Context context) {
        progressDialog = new ProgressDialog(context, R.style.MyTheme);
        progressDialog.setCancelable(false);
        progressDialog.setProgressStyle(android.R.style.Widget_ProgressBar_Small);
        progressDialog.show();
    }

    public static void stopProgressDialog(Context context) {
        if (progressDialog != null) {
            progressDialog.dismiss();
        }
    }

    public static boolean hasPermissionInManifest(Activity activity, int requestCode, String[] permissionName) {
        for (String s : permissionName) {
            if (ContextCompat.checkSelfPermission(activity, s) != PackageManager.PERMISSION_GRANTED) {
                // Some permissions are not granted, ask the user.
                ActivityCompat.requestPermissions(activity, permissionName, requestCode);
                return false;
            }
        }
        return true;
    }

    public static boolean isConnectingToInternet(Context context) {
        ConnectivityManager connectivity = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivity != null) {
            NetworkInfo[] info = connectivity.getAllNetworkInfo();
            if (info != null)
                for (int i = 0; i < info.length; i++)
                    if (info[i].getState() == NetworkInfo.State.CONNECTED) {
                        return true;
                    }
        }
        return false;
    }

    public static boolean isMyServiceRunning(Class<?> serviceClass, Context context) {
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                Log.e("servideRuning", "true");
                return true;
            }
        }
        return false;
    }


    public static void saveCountryList(Context context, List<CountryDetail> list) {
        String str = new Gson().toJson(list);
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(KEY_COUNTRY_LIST, str);
        editor.apply();
    }

    public static ArrayList<CountryDetail> getCountryList(Context context) {
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        String str = sp.getString(KEY_COUNTRY_LIST, null);
        try {
            CountryDetail[] arr = new Gson().fromJson(str, CountryDetail[].class);
            return new ArrayList<>(Arrays.asList(arr));
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public static void saveIndustryProfession(Context context, List<DetailIndustryProf> list) {
        String str = new Gson().toJson(list);
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(KEY_INDUSTRY_PROF, str);
        editor.apply();
    }

    public static ArrayList<DetailIndustryProf> getIndustryProfession(Context context) {
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        String str = sp.getString(KEY_INDUSTRY_PROF, null);
        try {
            DetailIndustryProf[] arr = new Gson().fromJson(str, DetailIndustryProf[].class);
            return new ArrayList<>(Arrays.asList(arr));
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public static void saveLocationHistory(Context context, List<DetailLocationHistory> list) {
        String str = new Gson().toJson(list);
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(KEY_LOCATION_HISTORY, str);
        editor.apply();
    }

    public static ArrayList<DetailLocationHistory> getLocationHistory(Context context) {
        SharedPreferences sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
        String str = sp.getString(KEY_LOCATION_HISTORY, null);
        try {
            DetailLocationHistory[] arr = new Gson().fromJson(str, DetailLocationHistory[].class);
            return new ArrayList<>(Arrays.asList(arr));
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    @SuppressLint("MissingPermission")
    public static ModelMyLocation getMyLocation(Context context) {
        ModelMyLocation modelMyLocation = new ModelMyLocation();
        FusedLocationProviderClient mFusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
        LocationRequest locationRequest = LocationRequest.create();
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
        mFusedLocationClient.getLastLocation().addOnCompleteListener(
                new OnCompleteListener<Location>() {
                    @Override
                    public void onComplete(@NonNull Task<Location> task) {
                        Location location = task.getResult();
                        double latitudes, longitudes, altitudes;
                        if (location != null) {
                             latitudes = location.getLatitude();
                             longitudes = location.getLongitude();
                             altitudes = location.getAltitude();

                            try {
                                Geocoder geocoder = new Geocoder(context, Locale.getDefault());
                                List<Address> addresses = geocoder.getFromLocation(latitudes, longitudes, 1);

                                String address = addresses.get(0).getAddressLine(0); // If any additional address line present than only, check with max available address lines by getMaxAddressLineIndex()
                                String city = addresses.get(0).getLocality();
                                String state = addresses.get(0).getAdminArea();
                                String country = addresses.get(0).getCountryName();
                                String postalCode = addresses.get(0).getPostalCode();
                                String knownName = addresses.get(0).getFeatureName(); // Only if available else return NULL

                                modelMyLocation.setLatitudes(location.getLatitude());
                                modelMyLocation.setLongitudes(location.getLongitude());
                                modelMyLocation.setAltitudes(location.getAltitude());
                                modelMyLocation.setAddress(address);
                                modelMyLocation.setCity(city);
                                modelMyLocation.setState(state);
                                modelMyLocation.setCountry(country);
                                modelMyLocation.setPostalCode(postalCode);
                                modelMyLocation.setKnownName(knownName);

                                Utility.setSharedPreference(context, Constants.KEY_LAT, String.valueOf(latitudes));
                                Utility.setSharedPreference(context, Constants.KEY_LONG, String.valueOf(longitudes));
                                Utility.setSharedPreference(context, Constants.KEY_MY_LOCATION, address);
//                                Log.e("modelMyLocationUTILITY", new Gson().toJson(modelMyLocation));
                            } catch (IOException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
        );
        return modelMyLocation;
    }

}

File: app/src/main/java/libertypassage/com/corporate/utilities/Utility.kt
package libertypassage.com.corporate.utilities

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityManager
import android.app.ProgressDialog
import android.content.Context
import android.content.pm.PackageManager
import android.location.Geocoder
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.gson.Gson
import libertypassage.com.corporate.R
import libertypassage.com.corporate.model.*
import java.io.IOException
import java.util.*
import java.util.regex.Pattern


object Utility {
    private const val PREFERENCE = "LIBERTY"
    private const val KEY_COUNTRY_LIST = "country_list"
    private const val KEY_INDUSTRY_LIST = "industry_list"
    private const val KEY_INDUSTRY_PROF = "industry_profession"
    private const val KEY_HOTSPOTS_LIST = "hotspots_list"
    private const val KEY_SAFE_ENTRY = "safe_entry"
    private const val KEY_VACCINE_LIST = "vaccine_list"
    private var progressDialog: ProgressDialog? = null


    fun setSharedPreference(context: Context, name: String?, value: String?) {
        val settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = settings.edit()
        editor.putString(name, value)
        editor.apply()
    }

    fun getSharedPreferences(context: Context, name: String?): String? {
        val settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        return settings.getString(name, "")
    }

    fun isEmailValid(email: String): Boolean {
        val expression = "^[\\w\\.-]+@([\\w\\-]+\\.)+[A-Z]{2,4}$"
        val pattern = Pattern.compile(expression, Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(email)
        return matcher.matches()
    }

    fun clearPreference(context: Context) {
        val settings = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = settings.edit()
        editor.clear()
        editor.apply()
    }

    fun showProgressDialog(context: Context?) {
        progressDialog = ProgressDialog(context, R.style.MyTheme)
        progressDialog!!.setCancelable(false)
        progressDialog!!.setProgressStyle(android.R.style.Widget_ProgressBar_Small)
        progressDialog!!.show()
    }

    fun stopProgressDialog(context: Context?) {
        if (progressDialog != null) {
            progressDialog!!.dismiss()
        }
    }

    fun hasPermissionInManifest(
            activity: Activity?,
            requestCode: Int,
            permissionName: Array<String?>
    ): Boolean {
        for (s in permissionName) {
            if (ContextCompat.checkSelfPermission(
                            activity!!,
                            s!!
                    ) != PackageManager.PERMISSION_GRANTED
            ) {
                // Some permissions are not granted, ask the user.
                ActivityCompat.requestPermissions(activity, permissionName, requestCode)
                return false
            }
        }
        return true
    }

    fun isConnectingToInternet(context: Context): Boolean {
        val connectivity = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val info = connectivity.allNetworkInfo
        for (i in info.indices) if (info[i].state == NetworkInfo.State.CONNECTED) {
            return true
        }
        return false
    }

    fun isMyServiceRunning(serviceClass: Class<*>, context: Context): Boolean {
        val manager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        for (service in manager.getRunningServices(Int.MAX_VALUE)) {
            if (serviceClass.name == service.service.className) {
                Log.e("servideRuning", "true")
                return true
            }
        }
        return false
    }

    fun saveIndustry(context: Context, list: List<DetailIndustry?>?) {
        val str = Gson().toJson(list)
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(KEY_INDUSTRY_LIST, str)
        editor.apply()
    }

    fun getIndustry(context: Context): ArrayList<DetailIndustry> {
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val str = sp.getString(KEY_INDUSTRY_LIST, null)
        return try {
            val arr = Gson().fromJson(str, Array<DetailIndustry>::class.java)
            return ArrayList<DetailIndustry>(Arrays.asList<DetailIndustry>(*arr))
        } catch (e: java.lang.Exception) {
            ArrayList()
        }
    }

    fun saveVaccine(context: Context, list: List<DetailVaccines?>?) {
        val str = Gson().toJson(list)
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(KEY_VACCINE_LIST, str)
        editor.apply()
    }

    fun getVaccine(context: Context): ArrayList<DetailVaccines> {
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val str = sp.getString(KEY_VACCINE_LIST, null)
        return try {
            val arr = Gson().fromJson(str, Array<DetailVaccines>::class.java)
            return ArrayList<DetailVaccines>(Arrays.asList<DetailVaccines>(*arr))
        } catch (e: java.lang.Exception) {
            ArrayList()
        }
    }


    fun saveIndustryProfession(context: Context, list: List<DetailIndustryProf?>?) {
        val str = Gson().toJson(list)
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(KEY_INDUSTRY_PROF, str)
        editor.apply()
    }

    fun getIndustryProfession(context: Context): ArrayList<DetailIndustryProf> {
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val str = sp.getString(KEY_INDUSTRY_PROF, null)
        return try {
            val arr = Gson().fromJson(str, Array<DetailIndustryProf>::class.java)
            return ArrayList<DetailIndustryProf>(Arrays.asList<DetailIndustryProf>(*arr))
        } catch (e: java.lang.Exception) {
            ArrayList()
        }
    }


    fun saveHotspots(context: Context, list: List<DetailHotSpots?>?) {
        val str = Gson().toJson(list)
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(KEY_HOTSPOTS_LIST, str)
        editor.apply()
    }

    fun getHotspots(context: Context): ArrayList<DetailHotSpots> {
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val str = sp.getString(KEY_HOTSPOTS_LIST, null)
        return try {
            val arr = Gson().fromJson(str, Array<DetailHotSpots>::class.java)
            return ArrayList<DetailHotSpots>(Arrays.asList<DetailHotSpots>(*arr))
        } catch (e: java.lang.Exception) {
            ArrayList()
        }
    }

    fun saveSafeEntry(context: Context, list: List<ModelSafeEntry?>?) {
        val str = Gson().toJson(list)
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val editor = sp.edit()
        editor.putString(KEY_SAFE_ENTRY, str)
        editor.apply()
    }

    fun getSafeEntry(context: Context): ArrayList<ModelSafeEntry> {
        val sp = context.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE)
        val str = sp.getString(KEY_SAFE_ENTRY, null)
        return try {
            val arr = Gson().fromJson(str, Array<ModelSafeEntry>::class.java)
            return ArrayList<ModelSafeEntry>(Arrays.asList<ModelSafeEntry>(*arr))
        } catch (e: java.lang.Exception) {
            ArrayList()
        }
    }

    fun getJsonFromAssets(context: Context, fileName: String?): String? {
        val jsonString: String
        jsonString = try {
            val `is` = context.assets.open(fileName!!)
            val size = `is`.available()
            val buffer = ByteArray(size)
            `is`.read(buffer)
            `is`.close()
            String(buffer)
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }
        return jsonString
    }

    @SuppressLint("MissingPermission")
    fun getMyLocation(context: Context?): ModelMyLocation? {
        val modelMyLocation =
            ModelMyLocation()
        val mFusedLocationClient = LocationServices.getFusedLocationProviderClient(
                context!!
        )
        val locationRequest = LocationRequest.create()
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        mFusedLocationClient.lastLocation.addOnCompleteListener { task ->
            val location = task.result
            val latitudes: Double
            val longitudes: Double
            val altitudes: Double
            if (location != null) {
                latitudes = location.latitude
                longitudes = location.longitude
                altitudes = location.altitude
                try {
                    val geocoder = Geocoder(context, Locale.getDefault())
                    val addresses = geocoder.getFromLocation(latitudes, longitudes, 1)
                    val address =
                            addresses[0].getAddressLine(0) // If any additional address line present than only, check with max available address lines by getMaxAddressLineIndex()
                    val city = addresses[0].locality
                    val state = addresses[0].adminArea
                    val country = addresses[0].countryName
                    val postalCode = addresses[0].postalCode
                    val knownName = addresses[0].featureName // Only if available else return NULL
                    modelMyLocation.latitudes = latitudes
                    modelMyLocation.longitudes = longitudes
                    modelMyLocation.altitudes = altitudes
                    modelMyLocation.address = address
                    modelMyLocation.city = city
                    modelMyLocation.state = state
                    modelMyLocation.country = country
                    modelMyLocation.postalCode = postalCode
                    modelMyLocation.knownName = knownName
                    Utility.setSharedPreference(
                            context,
                            Constants.KEY_LAT,
                            latitudes.toString()
                    )
                    Utility.setSharedPreference(
                            context,
                            Constants.KEY_LONG,
                            longitudes.toString()
                    )
                    setSharedPreference(
                            context,
                            Constants.KEY_MY_LOCATION,
                            address
                    )
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        return modelMyLocation
    }


}